File: viewer/decompiler/vm_candidate.py
"""VM bytecode 候选区段检测 — DEC3-D, 严守 §7.0 普适性.

设计目标: 把"巨大 OLLVM-VM 函数 (1056 块)"问题转化为
"VM bytecode 是这串字节, 你 (LLM) 推编码反汇编它".

普适性自查 (§7.0):
  ✓ 复用 viewer/ollvmdet.py heuristic 评分, 不假设特定 VM (AVMP/Themida/VMP/etc.)
  ✓ bytecode reader 检测靠通用 pattern (self-update load 高频出现), 不假设位宽
  ✓ 输出 hex dump 喂 LLM, **不 disasm** (LLM 推编码; 已有 mimo PoC 实证可行)
  ✓ ollvmdet 没检测到 → 没 candidate, 不影响其他功能 (backward compat)
  ✓ 没写死任何 SO 名 / fn 偏移 / opcode 编码 / 寄存器名

算法 (粗粒度):
  1. ollvm_detect_vm(trace) 拿可疑 dispatcher fn (置信度 + reasons)
  2. 在每个候选 fn 内, 找"高频 self-update load" (ldrh/ldrb/ldr with `!`):
     这是经典 VM bytecode reader 模式 — 单指令读字节并自增 IP
  3. 拿那条指令所有命中处的 base register 值序列, 取 min/max → bytecode 地址范围
  4. memshadow.hex_dump(min_addr, t=last_idx, rows=16) → 喂 LLM 用的 hex 视图
  5. 输出 VmCandidate, 不解码

Caveats (诚实标注):
  - 任何 hot self-update load 都会触发, 真假阳由 confidence + reasons 区分
  - LLM 看 hex dump 可推编码, 但 trace 短 (单 cmd) 时可能数据不够
  - bytecode 地址范围估值粗 — 极端情况 base reg 包绕也按 min/max 取
"""
from __future__ import annotations
from dataclasses import dataclass, field
from typing import Optional
import numpy as np
from ..disasm import decode


@dataclass
class VmCandidate:
    """One VM dispatcher 候选区."""
    dispatcher_pc: int                   # ollvmdet 给的 dispatcher 锚点
    confidence: float                    # ollvmdet 评分
    reasons: list[str] = field(default_factory=list)
    # bytecode reader: 高频 self-update load 指令
    reader_pc: int = 0
    reader_inst: str = ""                # mnemonic + op_str 文本 (LLM 看)
    reader_hits: int = 0                 # 该 PC 在 trace 命中次数
    reader_base_reg: str = ""            # 解析出的 base register
    # bytecode 字节范围 (mem 范围, 不是 record idx)
    bytecode_addr: int = 0
    bytecode_len: int = 0
    # hex dump (memshadow 抓的实际字节)
    hex_dump: list[str] = field(default_factory=list)


def _find_self_update_loads(trace, fn_idx_lo: int, fn_idx_hi: int,
                            min_hits: int = 8,
                            max_step: int = 16) -> list[tuple[int, int, str, str]]:
    """在 [fn_idx_lo, fn_idx_hi] 区间内找高频 self-update load.

    Returns: list of (pc, hits, mnem_op_str, base_reg_name).
    self-update 判定:
      - mnemonic in {ldrh, ldrb, ldr}
      - op_str 含 '!' (pre/post-update)
      - **|step| ≤ max_step** — VM bytecode reader 步长应该是单 opcode 大小
        (典型 1-8 字节, 极少 16+). 步长 #0x40 之类是 struct field access,
        过滤掉 (重要 false-positive 防线, 真机实证有效).

    Caveat: 这是 heuristic, 不是绝对. 个别 VM 用大步长 reader (跳过校验/填充
    字节) 会被误过滤. 真要看可调 max_step.
    """
    pc_arr = trace.pc_array()
    sub_pcs = pc_arr[fn_idx_lo:fn_idx_hi + 1]
    unique, counts = np.unique(sub_pcs, return_counts=True)
    order = np.argsort(-counts)
    hits_seen: dict[int, tuple[int, str, str]] = {}
    for i in order[:200]:
        pc = int(unique[i])
        cnt = int(counts[i])
        if cnt < min_hits:
            continue
        local_mask = sub_pcs == np.uint64(pc)
        if not local_mask.any():
            continue
        local_idx = int(np.argmax(local_mask)) + fn_idx_lo
        d = decode(pc, trace.inst(local_idx))
        if d.mnemonic not in ("ldrh", "ldrb", "ldr"):
            continue
        if "!" not in d.op_str:
            continue
        if not d.mem_op:
            continue
        base_reg, _idx_reg, disp, _sz, _is_w, _src = d.mem_op[0]
        # 步长 sanity 过滤
        if abs(int(disp)) > max_step:
            continue
        hits_seen[pc] = (cnt, f"{d.mnemonic} {d.op_str}", base_reg or "")
    return sorted(
        [(pc, c, ms, br) for pc, (c, ms, br) in hits_seen.items()],
        key=lambda x: -x[1],
    )


def _bytecode_range(trace, reader_pc: int, base_reg: str,
                    fn_idx_lo: int, fn_idx_hi: int) -> tuple[int, int]:
    """从所有命中 reader_pc 的 record 上拿 base_reg 值, 返回 (min, max).

    Returns (0, 0) 若解析失败 (无命中或 reg 名错).
    """
    if not base_reg:
        return (0, 0)
    pc_arr = trace.pc_array()
    sub_pcs = pc_arr[fn_idx_lo:fn_idx_hi + 1]
    mask = sub_pcs == np.uint64(reader_pc)
    if not mask.any():
        return (0, 0)
    idxs = np.nonzero(mask)[0] + fn_idx_lo
    vals = []
    for i in idxs[:5000]:                # cap to 5K hits, 够估范围
        try:
            r = trace.record(int(i))
            v = r.reg(base_reg)
            if v:
                vals.append(int(v))
        except Exception:
            continue
    if not vals:
        return (0, 0)
    return (min(vals), max(vals))


def detect_vm_candidates(trace, cfg, mem=None,
                         confidence_threshold: float = 0.4) -> list[VmCandidate]:
    """主入口: 从 trace 检测 VM dispatcher 候选 + 抓 bytecode hex.

    Args:
        trace: loaded Trace
        cfg: built CFG
        mem: optional MemShadow (建过的). None 则不出 hex_dump (省时间).
        confidence_threshold: ollvmdet 阈值

    Returns: list[VmCandidate], 按 confidence 降序.
    """
    from ..ollvmdet import ollvm_detect_vm
    findings = ollvm_detect_vm(trace, conf_threshold=confidence_threshold)
    if not findings:
        return []

    out: list[VmCandidate] = []
    n = len(trace)
    for f in findings:
        dispatcher_pc = int(f.get("fn_pc"), 16) if isinstance(f.get("fn_pc"), str) \
                        else int(f.get("fn_pc") or 0)
        cand = VmCandidate(
            dispatcher_pc=dispatcher_pc,
            confidence=float(f.get("confidence") or 0.0),
            reasons=str(f.get("reason", "")).split(" + ") if f.get("reason") else [],
        )
        # bytecode reader: 在整 trace 范围扫 (ollvmdet 没给 fn 边界, 用全程)
        readers = _find_self_update_loads(trace, 0, n - 1, min_hits=8)
        if readers:
            pc, hits, ms, base = readers[0]      # 最热的那条
            cand.reader_pc = pc
            cand.reader_inst = ms
            cand.reader_hits = hits
            cand.reader_base_reg = base
            lo, hi = _bytecode_range(trace, pc, base, 0, n - 1)
            if lo and hi and hi > lo:
                cand.bytecode_addr = lo
                # cap bytecode_len: base reg max-min 在地址跨段间会爆炸
                # (e.g. 跨 mmap region). 真实 VM bytecode 通常 < 64KB.
                # 超 64KB 标 'spans multiple regions, length unreliable'.
                raw_len = hi - lo + 1
                cand.bytecode_len = raw_len
                if mem is not None and mem.built:
                    # hex dump 只取 256 字节 (16 行 × 16), 足够 LLM 看 pattern
                    rows = 16
                    cand.hex_dump = mem.hex_dump(lo, n - 1, rows=rows, cols=16)
        out.append(cand)
    out.sort(key=lambda c: -c.confidence)
    return out

File: rust/crates/tracemiku-core/src/decompiler/vm_candidate.rs
//! VM bytecode 候选区段检测 (DEC3-D). Direct port of
//! viewer/decompiler/vm_candidate.py.
//!
//! Pipeline:
//!   1. ollvm_detect_vm(trace) → dispatcher candidates
//!   2. Find hottest self-update load (ldrh/ldrb/ldr with `!`) in trace
//!   3. min/max of base-reg values across all reader-PC hits → bytecode range
//!   4. memshadow.hex_dump(min_addr, last_idx, 16, 16) → LLM-readable hex
//!   5. Emit VmCandidateIR; never decode bytecode.

use crate::cfg::CFG;
use crate::decompiler::ir::VmCandidateIR;
use crate::disasm::decode;
use crate::index::Index;
use crate::memshadow::MemShadow;
use crate::ollvmdet::{ollvm_detect_vm, ollvm_detect_vm_indexed, OllvmFinding};
use crate::trace::Trace;

/// Find self-update loads in [lo, hi]: returns (pc, hits, mnem_op_str, base_reg).
/// Mirrors Python `_find_self_update_loads`.
fn find_self_update_loads(
    trace: &Trace,
    lo: usize,
    hi: usize,
    min_hits: u64,
    max_step: i64,
) -> Vec<(u64, u64, String, String)> {
    use std::collections::HashMap;

    if hi < lo || trace.is_empty() {
        return Vec::new();
    }

    // Frequency count over PCs in [lo, hi].
    let mut freq: HashMap<u64, u64> = HashMap::new();
    let cap = hi.min(trace.len() - 1);
    for i in lo..=cap {
        *freq.entry(trace.pc(i)).or_insert(0) += 1;
    }
    // Sort by hits desc; take top 200 (Python parity).
    let mut sorted: Vec<(u64, u64)> = freq.into_iter().collect();
    sorted.sort_by_key(|(_, c)| std::cmp::Reverse(*c));
    sorted.truncate(200);

    // For each candidate PC, decode its first-seen instance + apply filters.
    let mut hits_seen: Vec<(u64, u64, String, String)> = Vec::new();
    for (pc, cnt) in sorted {
        if cnt < min_hits {
            continue;
        }
        // Find first idx in [lo,hi] with this pc.
        let mut first_idx: Option<usize> = None;
        for i in lo..=cap {
            if trace.pc(i) == pc {
                first_idx = Some(i);
                break;
            }
        }
        let Some(idx) = first_idx else { continue };
        let inst = trace.inst(idx);
        let d = decode(pc, inst);
        let m = d.mnemonic.as_str();
        if m != "ldrh" && m != "ldrb" && m != "ldr" {
            continue;
        }
        if !d.op_str.contains('!') {
            continue;
        }
        let Some(mem_op) = d.mem_op.first() else {
            continue;
        };
        if mem_op.disp.unsigned_abs() as i64 > max_step {
            continue;
        }
        let mnem_op = format!("{} {}", d.mnemonic, d.op_str);
        hits_seen.push((pc, cnt, mnem_op, mem_op.base.clone()));
    }
    hits_seen.sort_by_key(|(_, c, _, _)| std::cmp::Reverse(*c));
    hits_seen
}

fn find_self_update_loads_indexed(
    trace: &Trace,
    index: &Index,
    lo: usize,
    hi: usize,
    min_hits: u64,
    max_step: i64,
) -> Vec<(u64, u64, String, String)> {
    if hi < lo || trace.is_empty() {
        return Vec::new();
    }
    let cap = hi.min(trace.len() - 1);
    let mut sorted: Vec<(u64, u64, usize)> = Vec::new();
    for (&pc, idxs) in &index.pc_to_idxs {
        let start = idxs.partition_point(|&idx| idx < lo);
        let end = idxs.partition_point(|&idx| idx <= cap);
        if start >= end {
            continue;
        }
        sorted.push((pc, (end - start) as u64, idxs[start]));
    }
    sorted.sort_by_key(|(_, c, _)| std::cmp::Reverse(*c));
    sorted.truncate(200);

    let mut hits_seen: Vec<(u64, u64, String, String)> = Vec::new();
    for (pc, cnt, idx) in sorted {
        if cnt < min_hits {
            continue;
        }
        let d = decode(pc, trace.inst(idx));
        let m = d.mnemonic.as_str();
        if m != "ldrh" && m != "ldrb" && m != "ldr" {
            continue;
        }
        if !d.op_str.contains('!') {
            continue;
        }
        let Some(mem_op) = d.mem_op.first() else {
            continue;
        };
        if mem_op.disp.unsigned_abs() as i64 > max_step {
            continue;
        }
        let mnem_op = format!("{} {}", d.mnemonic, d.op_str);
        hits_seen.push((pc, cnt, mnem_op, mem_op.base.clone()));
    }
    hits_seen.sort_by_key(|(_, c, _, _)| std::cmp::Reverse(*c));
    hits_seen
}

/// Walk all hits of `reader_pc` in [lo, hi]; pull `base_reg` value at each hit;
/// return (min, max). Returns (0, 0) on parse failure / no hits / unknown reg.
/// Mirrors Python `_bytecode_range`.
fn bytecode_range(
    trace: &Trace,
    reader_pc: u64,
    base_reg: &str,
    lo: usize,
    hi: usize,
) -> (u64, u64) {
    if base_reg.is_empty() {
        return (0, 0);
    }
    let n = trace.len();
    if n == 0 || hi < lo {
        return (0, 0);
    }
    let mut vals: Vec<u64> = Vec::new();
    let cap = hi.min(n - 1);
    for i in lo..=cap {
        if trace.pc(i) != reader_pc {
            continue;
        }
        let rec = trace.record(i);
        let Some(v) = rec.reg(base_reg) else { continue };
        if v != 0 {
            vals.push(v);
        }
        if vals.len() >= 5000 {
            break;
        }
    }
    if vals.is_empty() {
        return (0, 0);
    }
    let mn = *vals.iter().min().unwrap();
    let mx = *vals.iter().max().unwrap();
    (mn, mx)
}

fn bytecode_range_indexed(
    trace: &Trace,
    index: &Index,
    reader_pc: u64,
    base_reg: &str,
    lo: usize,
    hi: usize,
) -> (u64, u64) {
    if base_reg.is_empty() {
        return (0, 0);
    }
    let n = trace.len();
    if n == 0 || hi < lo {
        return (0, 0);
    }
    let Some(idxs) = index.pc_to_idxs.get(&reader_pc) else {
        return (0, 0);
    };
    let cap = hi.min(n - 1);
    let start = idxs.partition_point(|&idx| idx < lo);
    let end = idxs.partition_point(|&idx| idx <= cap);
    let mut vals: Vec<u64> = Vec::new();
    for &i in &idxs[start..end] {
        let rec = trace.record(i);
        let Some(v) = rec.reg(base_reg) else {
            continue;
        };
        if v != 0 {
            vals.push(v);
        }
        if vals.len() >= 5000 {
            break;
        }
    }
    if vals.is_empty() {
        return (0, 0);
    }
    let mn = *vals.iter().min().unwrap();
    let mx = *vals.iter().max().unwrap();
    (mn, mx)
}

/// Main entry: detect VM dispatcher candidates and grab bytecode hex.
///
/// `mem`: optional MemShadow (built). When `None`, emits candidates without
/// hex_dump (`hex_dump: vec![]`).
/// `confidence_threshold`: passed to ollvm_detect_vm.
pub fn detect_vm_candidates(
    trace: &Trace,
    cfg: &CFG,
    mem: Option<&MemShadow>,
    confidence_threshold: f64,
) -> Vec<VmCandidateIR> {
    detect_vm_candidates_inner(
        trace,
        cfg,
        mem,
        confidence_threshold,
        ollvm_detect_vm(trace, 10, confidence_threshold),
        find_self_update_loads,
        bytecode_range,
    )
}

pub fn detect_vm_candidates_indexed(
    trace: &Trace,
    cfg: &CFG,
    index: &Index,
    mem: Option<&MemShadow>,
    confidence_threshold: f64,
) -> Vec<VmCandidateIR> {
    detect_vm_candidates_inner(
        trace,
        cfg,
        mem,
        confidence_threshold,
        ollvm_detect_vm_indexed(trace, index, 10, confidence_threshold),
        |trace, lo, hi, min_hits, max_step| {
            find_self_update_loads_indexed(trace, index, lo, hi, min_hits, max_step)
        },
        |trace, reader_pc, base_reg, lo, hi| {
            bytecode_range_indexed(trace, index, reader_pc, base_reg, lo, hi)
        },
    )
}

fn detect_vm_candidates_inner<F, G>(
    trace: &Trace,
    _cfg: &CFG,
    mem: Option<&MemShadow>,
    _confidence_threshold: f64,
    findings: Vec<OllvmFinding>,
    find_readers: F,
    find_range: G,
) -> Vec<VmCandidateIR>
where
    F: Fn(&Trace, usize, usize, u64, i64) -> Vec<(u64, u64, String, String)>,
    G: Fn(&Trace, u64, &str, usize, usize) -> (u64, u64),
{
    if findings.is_empty() {
        return Vec::new();
    }
    let n = trace.len();
    if n == 0 {
        return Vec::new();
    }
    let mut out: Vec<VmCandidateIR> = Vec::new();
    for f in findings {
        let mut cand = VmCandidateIR {
            dispatcher_pc: f.fn_pc,
            confidence: f.confidence,
            reasons: f.reasons,
            ..Default::default()
        };
        let readers = find_readers(trace, 0, n - 1, 8, 16);
        if let Some((pc, hits, ms, base)) = readers.into_iter().next() {
            cand.reader_pc = pc;
            cand.reader_inst = ms;
            cand.reader_hits = hits;
            cand.reader_base_reg = base.clone();
            let (lo, hi) = find_range(trace, pc, &base, 0, n - 1);
            if lo > 0 && hi > lo {
                cand.bytecode_addr = lo;
                cand.bytecode_len = hi - lo + 1;
                if let Some(m) = mem {
                    cand.hex_dump = m.hex_dump(lo, (n - 1) as u64, 16, 16);
                }
            }
        }
        out.push(cand);
    }
    out.sort_by(|a, b| {
        b.confidence
            .partial_cmp(&a.confidence)
            .unwrap_or(std::cmp::Ordering::Equal)
    });
    out
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::cfg::build_cfg;
    use crate::index::Index;
    use crate::trace::REC_SIZE;

    fn synth_no_vm() -> tempfile::TempDir {
        // 12 nops — no indirect br → ollvm_detect_vm returns nothing.
        let dir = tempfile::tempdir().unwrap();
        let cd = dir.path().join("run").join("calls").join("c");
        std::fs::create_dir_all(&cd).unwrap();
        let mut buf = vec![0u8; REC_SIZE * 12];
        for i in 0..12usize {
            let off = i * REC_SIZE;
            buf[off..off + 8].copy_from_slice(&(0x1000u64 + (i as u64) * 4).to_le_bytes());
            buf[off + 268..off + 272].copy_from_slice(&0xd503201fu32.to_le_bytes());
        }
        std::fs::write(cd.join("trace.bin"), &buf).unwrap();
        std::fs::write(cd.join("meta.json"), r#"{"records":12}"#).unwrap();
        std::fs::write(
            dir.path().join("run").join("meta.json"),
            r#"{"module":{"name":"libt.so","base":"0x1000","size":"0x10000"}}"#,
        )
        .unwrap();
        dir
    }

    fn load_trace(dir: &tempfile::TempDir) -> Trace {
        let cd = dir
            .path()
            .join("run")
            .join("calls")
            .read_dir()
            .unwrap()
            .next()
            .unwrap()
            .unwrap()
            .path();
        Trace::load(&cd).unwrap()
    }

    #[test]
    fn detect_vm_candidates_empty_when_no_ollvm_signal() {
        let dir = synth_no_vm();
        let t = load_trace(&dir);
        let cfg = build_cfg(&t);
        let cands = detect_vm_candidates(&t, &cfg, None, 0.4);
        assert!(cands.is_empty(), "no ollvm signal → no candidates");
    }

    #[test]
    fn detect_vm_candidates_indexed_empty_when_no_ollvm_signal() {
        let dir = synth_no_vm();
        let t = load_trace(&dir);
        let cfg = build_cfg(&t);
        let index = Index::build(&t);
        let cands = detect_vm_candidates_indexed(&t, &cfg, &index, None, 0.4);
        assert!(cands.is_empty(), "no ollvm signal → no candidates");
    }

    #[test]
    fn find_self_update_loads_empty_on_no_match() {
        let dir = synth_no_vm();
        let t = load_trace(&dir);
        let n = t.len();
        let res = find_self_update_loads(&t, 0, n - 1, 1, 16);
        assert!(res.is_empty(), "no ldrh/ldrb/ldr with `!` in nops");
    }

    #[test]
    fn bytecode_range_zero_when_unknown_reg() {
        let dir = synth_no_vm();
        let t = load_trace(&dir);
        let n = t.len();
        let (lo, hi) = bytecode_range(&t, 0x9999, "x0", 0, n - 1);
        assert_eq!((lo, hi), (0, 0));
        let (lo, hi) = bytecode_range(&t, 0x1000, "", 0, n - 1);
        assert_eq!((lo, hi), (0, 0), "empty base_reg → (0,0)");
    }
}
